### PR TITLE
Nguyen/dev

### PR DESCRIPTION
**Pull request by nguyen-tran**
I finished fixing the database. RabbitMQ successfully pushes data right into the database whenever data is received. Also, I setup the telemetry data API so that the angular app can get the data by sending a GET request to /api/packet. Right now it returns literally the whole database on this request, but maybe we can include a time range in the payload to filter the results a bit.

Note: we will need to drop the existing tables in our DO and migrate the new tables I made in this repository. Otherwise, this app won't work on our droplet.

### CODE REVIEW

**Comment by enochtsang** [comment]
I think since this is consistent data, using websockets to emit data to clients would a better approach then exposing an API to get data once. What do you think?

**Comment by nguyen-tran** [comment]
The way I envisioned it is to stream or live-view the data through AJAX requests to the API. For example, using [chartkick](https://www.npmjs.com/package/chartkick) we give each chart its own endpoint and it automatically takes care of sending the AJAX request every n seconds and merges the new data into the chart without reloading the page, all in a single line of JavaScript. Or we can make our own AJAX request with timer and control how the data is displayed.

However, looking at [websockets](https://github.com/websockets/ws), I can see it wouldn't be that different from a RESTful API after the connection is established. I'm up for any method, as it should be a quick fix to change it now.

**Comment by enochtsang** [comment]
With simplicity in mind, I think websockets is still the better choice here. It's common javascript web design to have a single global state that everything pulls from. So in our case, this data is our state, and everything can pull from that state.

Websockets excel at pushing data to clients, and this is exactly what we're doing (and not really much else), so I think that is the technology we should be using to describe the API between our server and clients.

I know you already agreed with me to use websockets, but just to add some points and address some points that you had.

**Comment by nguyen-tran** [comment]
@enochtsang Okay, I've changed it to use websockets now from the 'ws' library. It still pretty much returns the entire database until we progress on the Angular app and decide how we want to structure our messages and payload.

File: migrate.txt [hunk 1]
```diff
@@ -0,0 +1,276 @@
+--
```
Comment (bill-luu) [inline migrate.txt]: The usual extension for these kinds of files is `.psql`.

File: server/README.adoc [hunk 2]
```diff
@@ -5,8 +5,40 @@ Express.js server running in node
...
+sudo su -l postgres
+....
+Next, create a database named epsilontelemetrydb +
+....
+createdb epsilontelemetrydb
+....
+Finally, import the schema (file is at the root of this repository) +
+....
+psql epsilontelemetrydb < migrate.txt
+....
```
Comment (bill-luu) [inline server/README.adoc]: We should mention a way to verify that it worked.
`\connect epsilontelmetrydb`
`\dt`

File: server/bin/www [hunk 3]
```diff
@@ -3,93 +3,45 @@
...
-var amqp = require('amqplib/callback_api');
-var io = require('socket.io');
+const app = require('../app');
+const db = require('../libs/database');
+const ws = require('../libs/websocket');
+
+const http = require('http');
+const amqp = require('amqplib/callback_api');
+
+const pg = require('pg');
```
Comment (bill-luu) [inline server/bin/www]: The [official doc](https://node-postgres.com/guides/project-structure) recommends to put the DB stuff in a separate file, so it's easier to track down as the project gets bigger.

Comment (nguyen-tran) [reply server/bin/www]: done, moved to startup scripts in libs/*.js

File: migrate.txt (hunk 1 above)
Comment (nguyen-tran) [reply migrate.txt]: done

File: server/libs/database.js [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+  mapping += jsonObj['MPPT'][2]['ArrayCurrent'] + ', ';
+  mapping += jsonObj['MPPT'][2]['BatteryVoltage'] + ', ';
+  mapping += jsonObj['MPPT'][2]['Temperature'] + ', ';
+  mapping += jsonObj['Lights']['LowBeams'] + ', ';
+  mapping += jsonObj['Lights']['HighBeams'] + ', ';
+  mapping += jsonObj['Lights']['Brakes'] + ', ';
+  mapping += jsonObj['Lights']['LeftSignal'] + ', ';
+  mapping += jsonObj['Lights']['RightSignal'] + ', ';
+  mapping += jsonObj['Lights']['BmsStrobeLight'] + ', ';
+  // Ignore lightsalive until Hermes is Updated
```
Comment (bill-luu) [inline server/libs/database.js]: Hermes has this value and the test tool will be updated with that tomorrow

File: server/README.adoc [hunk 5]
```diff
@@ -5,8 +5,45 @@ Express.js server running in node
 == Build
 `npm run start` to run dev server. Can be hit at `http://localhost:4000/`.
 
+== Setting up PostgreSQL
```
Comment (enochtsang) [inline server/README.adoc]: In the future, a docker container for this would be easiest.

File: server/libs/websocket.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ * Listens for a 'getData' event and sends the entire database back.
+ * 
+ * @param {server object from startup script} server 
+ */
+module.exports.socketSetupAndListen = function(server) {
+  const wss = new WebSocket.Server({server});
+
+  wss.on('connection', function(ws, req) {
+    // getData event handler
+    ws.on('getData', function(msg) {
```
Comment (enochtsang) [inline server/libs/websocket.js]: This is not really a different than a REST API and doesn't take advantage of websockets. The advantage websockets give us over a REST API is that the client doesn't need to ask, the server can just continuously give, but this doesn't really reflect that, the client shouldn't need to ask. The server should be broadcasting to all clients, all the time.

File: server/libs/websocket.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
+const WebSocket = require('ws');
+const db = require('./database');
+
+/**
+ * Function that handles setup and listening of websocket with the Angular app.
+ * Listens for a 'getData' event and sends the entire database back.
```
Comment (enochtsang) [inline server/libs/websocket.js]: Not the whole database right? The whole packet table.

File: server/app.js [hunk 8]
```diff
@@ -1,27 +1,23 @@
...
-app.use('/', index);
+app.use(logger('dev'));
+app.use(express.static(path.join(__dirname, '../web-app/dist')));
+
+/**
+ * Routes
+ */
+// Main page
+app.use('/', (req, res) => {
+  res.sendFile(path.join(__dirname, '../web-app/dist/index.html'));
```
[the remark this answers is not in the document]
Comment (enochtsang) [reply server/app.js]: So forgive my ignorance of angular, how does angular know how to generate that file? At this time, I see no instructions on how to generate the frontend, even a simple one.

[the remark this answers is not in the document]
Comment (nguyen-tran) [reply server/app.js]: Angular will compile any changes made to the source when you start with `ng serve`.

**Comment by nguyen-tran** [comment]
I've been reading up on the Websocket library and general websocket standards. I believe it's best to keep sending plain text between the Node and Angular app with the current setup.

If we really wanted to be sending JSON (or any object), we would have to first serialize it and send it as binary data. This would introduce additional overhead on the client side that isn't really beneficial for us since everything is hosted on the same localhost anyways.

File: server/app.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (enochtsang) [reply server/app.js]: I don't mean instrucitons on how to generate the frontend, but I see no HTML or JSX anywhere. 

EDIT: Now I see that it's there in the repo, just not part of this PR.

File: server/libs/websocket.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        }
+
+        // send latest telemetry row to client (array with 1 element)
+        wss.send(bson.serialize(result.rows));
+    });
+  });
+
+  // get_past_data event handler to view historical telemetry data
+  // MUST receive a JSON object with fields 'start' and 'end' with timestamps
+  ws.on('get_past_data', function(msg) {
```
Comment (enochtsang) [inline server/libs/websocket.js]: Im not 100% what the intended use for this for, but this shouldn't be the normal use case. The server should be keeping a list of all the clients and continuously sending the data to it. On connection, it should just be putting it in the connection in a list, and some other task is sending the current data to all clients every second or so.

File: server/libs/amqp.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+            console.log('1 row inserted from RabbitMQ');
+
+            // serialize the data in the proper format
+            let serData = [];
+            serData.push(result.rows[0].row_to_json)
+            serData = bson.serialize(serData);
+
+            // broadcast inserted row to all connected clients
+            wss.broadcast = function broadcast(serData) {
+              wss.clients.forEach(function each(client) {
```
Comment (enochtsang) [inline server/libs/amqp.js]: I would use the `map()` function, it's more of a functional approach, can be easier to reason through due to less possible side effects, and also functional is the future in web dev ;)

File: server/libs/amqp.js [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+});
+console.log('Column names updated');
+
+/**
+ * Setup the AMQP channel with RabbitMQ (change host if needed)
+ */
+const rmq_URL = 'amqp://localhost:5672';
+amqp.connect(rmq_URL, function(err, conn) {
+  try {
+    conn.createChannel(function(err, ch) {
```
Comment (enochtsang) [inline server/libs/amqp.js]: Are promises supported? That would look nicer compared to having a huge try catch block.

File: server/libs/amqp.js [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+      // exchange name must be hermesExchange
+      const q = 'msgQueue';
```
Comment (enochtsang) [inline server/libs/amqp.js]: We should read these from a config someday, but it's a TODO @bill-luu

File: server/libs/amqp.js [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+
+        // connect to database and check for errors
+        db.pool.connect(function(err, client, done) {
+          if (err) {
+            return console.error(db.errors.CONNECT_ERROR, err.stack);
+          }
+
+          // insert into database and return inserted row as JSON
+          const query = 'INSERT INTO packet(' + columnNames + ') ' +
+                        'values(' + db.mapJsonToColumns(jsonObj) + ') ' +
```
Comment (enochtsang) [inline server/libs/amqp.js]: This looks like it's vulnerable to SQL injection, is it possible to use a prepared statement? Or possibly the library `knex`?

File: server/libs/websocket.js (hunk 9 above)
Comment (enochtsang) [reply server/libs/websocket.js]: EDIT: Disregard, this comment is not relevant. Nguyen's design is good.

File: server/libs/amqp.js (hunk 13 above)
Comment (nguyen-tran) [reply server/libs/amqp.js]: One of the beauties of using PostgreSQL, the postgres server escapes all strings for us :)

https://github.com/brianc/node-postgres/wiki/FAQ 8th point

File: server/libs/database.js [hunk 14]
```diff
@@ -188,12 +185,10 @@ module.exports.mapJsonToColumns = function(jsonObj) {
   mapping += jsonObj['Battery']['HighCellVoltage'] + ', ';
   mapping += jsonObj['Battery']['HighCellVoltageId'] + ', ';
   mapping += jsonObj['Battery']['AverageCellVoltage'] + ', ';
-  // Ignore PrechargeState until Hermes is updated
-  // and replace with Off instead
-  // mapping += jsonObj['Battery']['PrechargeState'] + ', ';
+  // mapping += jsonObj['AuxBms']['PrechargeState'] + ', ';
```
Comment (bill-luu) [inline server/libs/database.js]: Does this still need to be fixed?

Comment (nguyen-tran) [reply server/libs/database.js]: I just pulled the latest changes for the test tool and they're still sending the wrong values. The documentation says that it must be one of: `'Off', 'Common Engaged', 'Charge Engaged', 'Discharge Engaged', 'All Engaged'`.

Currently it is sending `'Enable Pack'`

File: server/config.json [hunk 15]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (enochtsang) [inline server/config.json]: We should make this file just an example file. And not version control the actual config file.

**Comment by bill-luu** [comment]
Tested and it looks good! Great job with this Nguyen :)

**Pull request merged**